expression/collation.go

```diff
@@ -184,39 +182,51 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
-	curCoer := CoercibilityIgnorable
-	curCollationPriority := 0
+	dstCollation, dstCharset, _, _ = inferCollation(exprs...)
+	return
+}
+
+// inferCollation infers collation, charset, coercibility and check the legitimacy.
+func inferCollation(exprs ...Expression) (dstCollation, dstCharset string, coercibility Coercibility, legal bool) {
+	firstExplicitCollation := ""
+	coercibility = CoercibilityIgnorable
 	dstCharset, dstCollation = charset.GetDefaultCharsetAndCollate()
-	if ctx != nil && ctx.GetSessionVars() != nil {
-		dstCharset, dstCollation = ctx.GetSessionVars().GetCharsetInfo()
-		if dstCharset == "" || dstCollation == "" {
-			dstCharset, dstCollation = charset.GetDefaultCharsetAndCollate()
-		}
-	}
-	hasStrArg := false
-	// see https://dev.mysql.com/doc/refman/8.0/en/charset-collation-coercibility.html
-	for _, e := range exprs {
-		if e.GetType().EvalType() != types.ETString {
-			continue
-		}
-		hasStrArg = true
-
-		coer := e.Coercibility()
-		ft := e.GetType()
-		priority := collationPriority[strings.ToLower(ft.Collate)]
-		if coer != curCoer {
-			if coer < curCoer {
-				curCoer, curCollationPriority, dstCharset, dstCollation = coer, priority, ft.Charset, ft.Collate
+	for _, arg := range exprs {
+		if arg.Coercibility() == CoercibilityExplicit {
+			if firstExplicitCollation == "" {
+				firstExplicitCollation = arg.GetType().Collate
+				coercibility, dstCollation, dstCharset = CoercibilityExplicit, arg.GetType().Collate, arg.GetType().Charset
+			} else if firstExplicitCollation != arg.GetType().Collate {
+				return "", "", CoercibilityIgnorable, false
+			}
+		} else if arg.Coercibility() < coercibility {
+			coercibility, dstCollation, dstCharset = arg.Coercibility(), arg.GetType().Collate, arg.GetType().Charset
+		} else if arg.Coercibility() == coercibility && dstCollation != arg.GetType().Collate {
+			p1 := collationPriority[dstCollation]
+			p2 := collationPriority[arg.GetType().Collate]
+
+			// same priority means this two collation is incompatible, coercibility might derive to CoercibilityNone
+			if p1 == p2 {
+				coercibility, dstCollation, dstCharset = CoercibilityNone, getBinCollation(arg.GetType().Charset), arg.GetType().Charset
+			} else if p1 < p2 {
+				dstCollation, dstCharset = arg.GetType().Collate, arg.GetType().Charset
 			}
-			continue
-		}
-		if priority <= curCollationPriority {
-			continue
 		}
-		curCollationPriority, dstCharset, dstCollation = priority, ft.Charset, ft.Collate
 	}
-	if !hasStrArg {
-		dstCharset, dstCollation = charset.CharsetBin, charset.CollationBin
+
+	return dstCollation, dstCharset, coercibility, true
+}
+
+// getBinCollation get binary collation by charset
+func getBinCollation(cs string) string {
+	switch cs {
+	case charset.CharsetUTF8:
+		return charset.CollationUTF8
+	case charset.CharsetUTF8MB4:
+		return charset.CollationUTF8MB4
 	}
-	return
+
+	logutil.BgLogger().Error("unexpected charset" + cs)
+	// it must return something, never reachable
+	return ""
```

Comment:
I think it'ok to fall back to charset.CollationUTF8MB4, returning "" is not safe.